src/uct/ib/mlx5/ib_mlx5.h

```diff
@@ -564,6 +567,72 @@ ucs_status_t uct_ib_mlx5_devx_modify_qp_state(uct_ib_mlx5_qp_t *qp,
 
 void uct_ib_mlx5_devx_destroy_qp(uct_ib_mlx5_qp_t *qp);
 
+enum {
+    UCT_IB_MLX5_ALLOC_FLAG_DONTFORK = UCS_BIT(0),
+    UCT_IB_MLX5_ALLOC_FLAG_SILENT   = UCS_BIT(1),
+};
+
+static inline ucs_status_t
+uct_ib_mlx5_buf_alloc(void **buf_p, uct_ib_mlx5_devx_umem_t *mem, size_t len,
+                      uct_ib_mlx5_md_t *md, int silent, char *name)
+{
+    ucs_log_level_t level = silent ? UCS_LOG_LEVEL_DEBUG : UCS_LOG_LEVEL_ERROR;
+    ucs_status_t status;
+    void *buf;
+    int ret;
+
+    ret = ucs_posix_memalign(&buf, ucs_get_page_size(), len, name);
+    if (ret != 0) {
+        ucs_log(level, "failed to allocate QP buffer of %zu bytes: %m", len);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (md->super.fork_init) {
+        ret = madvise(buf, len, MADV_DONTFORK);
+        if (ret != 0) {
+            ucs_log(level, "madvise(DONTFORK, buf=%p, len=%zu) failed: %m", buf, len);
+            status = UCS_ERR_IO_ERROR;
+            goto err_free;
+        }
+        mem->size = len;
+    } else {
+        mem->size = 0;
```

Comment:
seems a bit hacky to set size=0 to avoid madvise
IMO better pass MD to uct_ib_mlx5_buf_free

Reply:
IMO we could set size=UCT_IB_MD_NO_MADVISE(=0) and check it in free. This way it both wouldn't look hacky and spare us from MD propagation in destroy flow